File: src/job_matching.rs
use std::fs::File;
use std::io::Read;
use std::collections::HashMap;
use std::vec::Vec;

use csv::ReaderBuilder;
use ndarray::{Array1, Array2};
use ndarray_linalg::norm::Norm;
use ndarray_stats::QuantileExt;
use serde::{Deserialize, Serialize};
use nlp_units::{tfidf_vectorize, cosine_similarity};

#[derive(Debug, Deserialize, Serialize)]
struct JobPosting {
    date: String,
    Title: String,
    Company: String,
    Eligibility: String,
    Location: String,
    JobDescription: String,
    JobRequirment: String,
    RequiredQual: String,
    ApplicationP: String,
}

struct JobMatching {
    jobs: Vec<JobPosting>,
}

impl JobMatching {
    fn new() -> JobMatching {
        let jobs = JobMatching::load_jobs().unwrap();
        JobMatching {
            jobs: jobs.iter().take(500).cloned().collect(),
        }
    }

    fn load_resume(&self) -> String {
        let mut file = File::open("data/Resume-Iqra-2023.txt").unwrap();
        let mut contents = String::new();
        file.read_to_string(&mut contents).unwrap();
        contents
    }

    fn load_jobs() -> Result<Vec<JobPosting>, csv::Error> {
        let file = File::open("data/job_posts.csv").unwrap();
        let mut reader = ReaderBuilder::new()
            .has_headers(true)
            .delimiter(b',')
            .flexible(true)
            .from_reader(file);

        let mut jobs = Vec::new();

        for result in reader.deserialize() {
            let record: JobPosting = result?;
            if !record.JobDescription.is_empty() && !record.JobRequirment.is_empty() {
                jobs.push(record);
            }
        }

        Ok(jobs)
    }

    fn get_all_jobs(&self) -> Vec<HashMap<String, String>> {
        let mut jobs = Vec::new();

        for job in &self.jobs {
            let mut job_map = HashMap::new();
            job_map.insert("date".to_string(), job.date.clone());
            job_map.insert("Title".to_string(), job.Title.clone());
            job_map.insert("Company".to_string(), job.Company.clone());
            job_map.insert("Eligibility".to_string(), job.Eligibility.clone());
            job_map.insert("Location".to_string(), job.Location.clone());
            job_map.insert("JobDescription".to_string(), job.JobDescription.clone());
            job_map.insert("JobRequirment".to_string(), job.JobRequirment.clone());
            job_map.insert("RequiredQual".to_string(), job.RequiredQual.clone());
            job_map.insert("ApplicationP".to_string(), job.ApplicationP.clone());

            jobs.push(job_map);
        }

        jobs
    }


    fn keyword_matching(resume: &str, jobs: &mut Vec<HashMap<String, String>>) {
        let mut vectorizer = CountVectorizer::new();
        let keywords = vectorizer.fit_transform(vec![resume]).unwrap().toarray().unwrap().remove(0);

        let mut keywords_match = vec![];
        for i in 0..jobs.len() {
            let job_description_keywords = vectorizer.transform(vec![&jobs[i]["JobDescription"]]).unwrap().toarray().unwrap().remove(0);
            let job_requirement_keywords = vectorizer.transform(vec![&jobs[i]["JobRequirment"]]).unwrap().toarray().unwrap().remove(0);
            let mut match_found = true;
            for j in 0..keywords.len() {
                if job_description_keywords[j] < keywords[j] && job_requirement_keywords[j] < keywords[j] {
                    match_found = false;
                    break;
                }
            }
            keywords_match.push(match_found);
        }

        // Add keywords_match column to jobs vector of hashmaps
        for i in 0..jobs.len() {
            jobs[i].insert("keywords_match".to_string(), keywords_match[i].to_string());
        }
    }


    fn cosine_similarity(&mut self, resume: &str, jobs: &mut DataFrame) {
        let vectorizer = TfidfVectorizer::default();
        let resume_vector = vectorizer.fit_transform(&[resume]);

        let mut similarity_scores = Vec::new();
        for i in 0..jobs.height() {
            let job_vector = vectorizer.transform(&[jobs["jobpost"][i].clone()]);
            let similarity = cosine_similarity(&resume_vector, &job_vector)[0][0];
            similarity_scores.push(similarity);
        }

        jobs.add_column("cosine_similarity".to_string(), Series::new(similarity_scores));
    }

    fn get_jobs_matched(&self, resume: &str) -> Vec<HashMap<String, String>> {
        let mut jobs = self.jobs.clone();
        self.cosine_similarity(resume, &mut jobs);
        jobs.sort_by(|a, b| b["cosine_similarity"].partial_cmp(&a["cosine_similarity"]).unwrap());
        jobs = jobs.iter().map(|job| {
            let mut filtered_job = HashMap::new();
            filtered_job.insert("date".to_string(), job["date"].clone());
            filtered_job.insert("Title".to_string(), job["Title"].clone());
            filtered_job.insert("Company".to_string(), job["Company"].clone());
            filtered_job.insert("Eligibility".to_string(), job["Eligibility"].clone());
            filtered_job.insert("Location".to_string(), job["Location"].clone());
            filtered_job.insert("JobDescription".to_string(), job["JobDescription"].clone());
            filtered_job.insert("JobRequirment".to_string(), job["JobRequirment"].clone());
            filtered_job.insert("RequiredQual".to_string(), job["RequiredQual"].clone());
            filtered_job.insert("ApplicationP".to_string(), job["ApplicationP"].clone());
            filtered_job
        }).collect();
        jobs.truncate(20);
        jobs
    }
}

fn main() {
    let mut job_matching = JobMatching::new();
    let resume = job_matching.load_resume();
    let results = job_matching.get_jobs_matched(&resume);
    println!("{:?}", results);
}
    

File: job_matching.py
import pandas as pd
import numpy as np
from sklearn.feature_extraction.text import CountVectorizer
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.metrics.pairwise import cosine_similarity
import sys
sys.path.append('.')
#from gensim.models import Word2Vec
#from gensim.models.fasttext import FastText
#from keras.preprocessing.text import Tokenizer
#from tensorflow.keras.preprocessing.sequence import pad_sequences
#from keras.layers import Input, Embedding, LSTM, Dense, Lambda
#from keras.models import Model
#from keras import backend as K
#import gensim.downloader as api

class JobMatching:
    def __init__(self):
        self.jobs = self.load_jobs()

    
    def load_resume(self):
        resume = open("data/Resume-Iqra-2023.txt").read()
        return resume
    
    def load_jobs(self):
        jobs = pd.read_csv("data/dataworld-jobs-sample.csv")
        jobs.dropna(subset=['Job Description'], inplace=True)
        jobs.fillna('', inplace=True)
        jobs = jobs.rename(columns={'Job Title': 'JobTitle', 'Job Description': 'JobDescription', 'Job Type':'JobType'})
        return jobs
    
    def get_all_jobs(self):
        jobs = self.jobs
        jobs = jobs[['JobTitle', 'JobDescription', 'JobType', 'Categories',
       'Location', 'City', 'State', 'Country', 'Zip Code', 'Address',
       'Salary From', 'Salary To', 'Salary Period', 'Apply Url', 'Apply Email',
       'Employees', 'Industry', 'Company Name', 'Employer Email',
       'Employer Website', 'Employer Phone', 'Employer Logo',
       'Companydescription', 'Employer Location', 'Uniq Id',
       'Crawl Timestamp']]
        return jobs.head(20).to_dict(orient='records')
    
    def keyword_matching(self,resume,jobs):
        vectorizer = CountVectorizer()
        keywords = vectorizer.fit_transform([resume]).toarray()[0]

        keywords_match = []
        for i in range(len(jobs)):
            job_description_keywords = vectorizer.transform([jobs['JobDescription'].iloc[i]]).toarray()[0]
            job_requirement_keywords = vectorizer.transform([jobs['JobRequirment'].iloc[i]]).toarray()[0]
            match = all(job_description_keywords[i] >= keywords[i] or job_requirement_keywords[i] >= keywords[i] for i in range(len(keywords)))
            keywords_match.append(match)

        # Add keywords_match column to books dataframe
        jobs['keywords_match'] = keywords_match
        

    '''def siamese_similarity(self, text1, text2):
        # Tokenize the texts
        tokenizer = Tokenizer()
        tokenizer.fit_on_texts([text1, text2])

        # Convert the texts to sequences of indices
        seq1 = tokenizer.texts_to_sequences([text1])[0]
        seq2 = tokenizer.texts_to_sequences([text2])[0]

        # Pad the sequences to a fixed length
        max_len = 20
        padded_seq1 = pad_sequences([seq1], maxlen=max_len, padding='post', truncating='post')[0]
        padded_seq2 = pad_sequences([seq2], maxlen=max_len, padding='post', truncating='post')[0]

        # Load the pre-trained FastText model
        #model = FastText.load_fasttext_format('cc.en.300.bin')
        model = api.load('fasttext-wiki-news-subwords-300')

        # Encode the sequences using the FastText model
        embedding_matrix = np.zeros((len(tokenizer.word_index) + 1, model.vector_size))
        for word, i in tokenizer.word_index.items():
            if word in model:
                embedding_matrix[i] = model[word]

        left_input = Input(shape=(max_len,))
        right_input = Input(shape=(max_len,))

        embedding_layer = Embedding(input_dim=len(tokenizer.word_index) + 1,
                                    output_dim=model.vector_size,
                                    weights=[embedding_matrix],
                                    input_length=max_len,
                                    trainable=False)

        left_embedded = embedding_layer(left_input)
        right_embedded = embedding_layer(right_input)

        lstm_units = 64
        lstm_layer = LSTM(lstm_units)

        left_lstm = lstm_layer(left_embedded)
        right_lstm = lstm_layer(right_embedded)

        # Define the similarity function (in this case, cosine similarity)
        def cosine_similarity(x):
            left, right = x
            dot_product = left * right
            dot_product = K.sum(dot_product, axis=1, keepdims=True)
            left_norm = K.sqrt(K.sum(K.square(left), axis=1, keepdims=True))
            right_norm = K.sqrt(K.sum(K.square(right), axis=1, keepdims=True))
            return dot_product / (left_norm * right_norm)

        similarity_layer = Lambda(cosine_similarity)

        similarity = similarity_layer([left_lstm, right_lstm])

        siamese_model = Model(inputs=[left_input, right_input], outputs=similarity)

        siamese_model.compile(loss='binary_crossentropy', optimizer='adam', metrics=['accuracy'])

        # Predict the similarity between the two texts
        similarity_score = siamese_model.predict([padded_seq1.reshape(1, -1), padded_seq2.reshape(1, -1)])[0][0]
        return similarity_score

    
    def siamese_scores(self,resume,jobs):
        similarity_scores = []
        for i in range(len(jobs)):
            similarity = self.siamese_similarity(resume, jobs['JobDescription'].iloc[i])
            similarity_scores.append(similarity)
            
        jobs['siamese_similarity'] = similarity_scores
        jobs = jobs.sort_values('siamese_similarity', ascending=False)'''
            
    def cosine_similarity(self,resume,jobs):
        vectorizer = TfidfVectorizer()
        resume_vector = vectorizer.fit_transform([resume])

        similarity_scores = []
        for i in range(len(jobs)):
            job_vector = vectorizer.transform([jobs['JobDescription'].iloc[i]])
            similarity = cosine_similarity(resume_vector, job_vector)[0][0]
            similarity_scores.append(similarity)
            
        jobs['cosine_similarity'] = similarity_scores
        
    def get_jobs_matched(self,resume):
        jobs = self.jobs
        self.cosine_similarity(resume,jobs)
        #self.siamese_scores(resume,jobs)
        jobs = jobs.sort_values(by=['cosine_similarity'], ascending=False)
        #jobs = jobs[['date','Title','Company','Eligibility','Location','JobDescription','JobRequirment','RequiredQual','ApplicationP']]
        return jobs.head(20).to_dict(orient='records')
        
if __name__ == '__main__':
    job_matching = JobMatching()
    resume = job_matching.load_resume()
    results = job_matching.get_jobs_matched(resume)
    print(results)
    

        